io.home-assistant/index.js

```diff
@@ -0,0 +1,279 @@
+    constructor(engine, state, master, entityId) {
+        super(engine, state);
+
+        this.master = master;
+        this._entityId = entityId;
+        this.uniqueId = master.uniqueId + '/' + entityId;
+        this.name = this.state.friendly_name;
+    }
+
+    updateState(state) {
+        super.updateState(state);
+        this.name = this.state.friendly_name;
+    }
+
+    _callService(domain, service, data = {}) {
+        data.entity_id = this._entityId;
+        return HomeAssistant.callService(this.master.connection, domain, service, data);
+    }
+}
+
+class HomeAssistantLightbulbDevice extends HomeAssistantDevice {
+    constructor(engine, state, master, entityId) {
+        super(engine, state, master, entityId);
+
+        this.uniqueId = master.uniqueId + '/' + entityId;
+    }
+
+    async do_set_power({ power }) {
+        if (power === 'on')
+            await this._callService('light', 'turn_on');
+        else
+            await this._callService('light', 'turn_off');
+    }
+
+    async do_alert_long() {
+        await this._callService('light', 'turn_on', { flash: 'long' });
+    }
+
+    async do_color_loop() {
+        await this._callService('light', 'turn_on', { effect: 'colorloop' });
+    }
+}
+
+class HomeAssistantDeviceSet extends Tp.Helpers.ObjectSet.Base {
+    constructor(master) {
+        super();
+        this.master = master;
+        this._devices = new Map;
+
+        this._unsubscribe = null;
+    }
+
+    _maybeAddEntity(entityId, entityData) {
+        const existing = this._devices.get(entityId);
+        if (existing) {
+            // add "kind" property that Tp expects
+            entityData.kind = existing.state.kind;
+            existing.updateState(entityData);
+            return;
+        }
+
+        const [type,] = entityId.split('.');
+
+        let device;
+        switch (type) {
+        case 'light':
+            entityData.kind = 'light-bulb';
+            device = new HomeAssistantLightbulbDevice(this.engine, entityData, this.master, entityId);
+            break;
+        default:
+            console.log(`Unhandled Home Assistant entity ${entityId} of type ${type}`);
+            return;
+        }
+
+        this._devices.set(entityId, device);
+        this.objectAdded(device);
+    }
+
+    _maybeDeleteEntity(entityId) {
+        const device = this._devices.get(entityId);
+        if (!device)
+            return;
+        this._devices.delete(entityId);
+        this.objectRemoved(device);
+    }
+
+    async start() {
+        const existing = await HomeAssistant.getStates(this.master.connection);
+
+        for (let entity of existing)
+            this._maybeAddEntity(entity.entity_id, entity.attributes);
+
+        this._unsubscribe = this.master.connection.subscribeEvents((event) => {
+            const { entity_id, new_state } = event.data;
+            if (new_state)
+                this._maybeAddEntity(entity_id, new_state.attributes);
```

Comment:
By only passing the attributes, you're missing the most important part: the state. You should also pass `state.state`. That's how you know if a light is on.